graylog2-web-interface/src/components/configurations/SearchesConfig.jsx

```diff
@@ -0,0 +1,288 @@
+    };
+  },
+
+  _getPropConfigValue(field) {
+    return this.props.config ? this.props.config[field] : null;
+  },
+
+  // Converts the object to a list of objects to make form handling easier.
+  _createTimerangeOptionsState(options) {
+    let optionsList = null;
+
+    if (options) {
+      optionsList = Object.keys(options).map((key) => {
+        return {period: key, description: options[key]};
+      });
+    }
+
+    return optionsList;
+  },
+
+  _updateTimerangeOptionsState(options) {
+    const config = this.state.config;
+    const newTimerangeOptions = {};
+
+    options.forEach((option) => {
+      newTimerangeOptions[option.period] = option.description;
+    });
+
+    config.relative_timerange_options = newTimerangeOptions;
+
+    this.setState({config: config, timerangeOptionsList: options});
+  },
+
+  _isValidPeriod(duration) {
+    const check = duration || this.state.config.query_time_range_limit;
+    return moment.duration(check).asMilliseconds() >= 0;
+  },
+
+  _validationState() {
+    if (this._isValidPeriod()) {
+      return undefined;
+    } else {
+      return 'error';
+    }
+  },
+
+  _formatDuration(duration) {
+    return this._isValidPeriod() ? moment.duration(duration).humanize() : 'invalid';
+  },
+
+  _onPeriodUpdate(field) {
+    return () => {
+      const update = this.state.config;
+      let period = this.refs[field].getValue().toUpperCase();
+
+      if (!period.startsWith('P')) {
+        period = `P${period}`;
+      }
+
+      update[field] = period;
+
+      this.setState({config: update});
+    };
+  },
+
+  _onChecked() {
+    const config = this.state.config;
+
+    if (this.state.limitEnabled) {
+      // If currently enabled, disable by setting the limit to 0 seconds.
+      config.query_time_range_limit = 'PT0S';
+    } else {
+      // If currently not enabled, set a default of 30 days.
+      config.query_time_range_limit = 'P30D';
+    }
+
+    this.setState({config: config, limitEnabled: !this.state.limitEnabled});
+  },
+
+  _onTimerangeOptionRemove(removedIdx) {
+    return () => {
+      const options = this.state.timerangeOptionsList;
+
+      // Remove element at index
+      options.splice(removedIdx, 1);
+
+      this._updateTimerangeOptionsState(options);
+    };
+  },
+
+  _onTimerangeOptionAdd() {
+    const options = this.state.timerangeOptionsList;
+
+    if (options) {
+      options.push({period: 'PT1S', description: ''});
+      this._updateTimerangeOptionsState(options);
+    }
+  },
+
+  _onTimerangeOptionChange(changedIdx, field) {
+    return (e) => {
+      const options = this.state.timerangeOptionsList;
+
+      options.forEach((o, idx) => {
+        if (idx === changedIdx) {
+          let value = e.target.value;
+
+          if (field === 'period') {
+            value = value.toUpperCase();
+            if (!value.startsWith('P')) {
+              value = `P${value}`;
+            }
+          }
+
+          options[idx][field] = value;
```

Comment:
Here we are mutating the component's state directly, and this is highly discourage in the documentation: https://facebook.github.io/react/docs/component-api.html#setstate. You can use `ObjectUtils.clone` to clone the state object before modifying it.